web/pdf_find_controller.js

```diff
@@ -108,11 +112,67 @@ var PDFFindController = (function PDFFindControllerClosure() {
       });
     },
 
+    // Helper for multiple search.
+    // Sorting array of objects { match: <match>, matchLength: <matchLength> }
+    // in increasing index first and then the lengths.
+    sortMatches: function PDFFindController_sortMatches(matchesWithLength) {
+      matchesWithLength.sort(function(a, b) {
+        return a.match === b.match ?
+          a.matchLength - b.matchLength : a.match - b.match;
+        });
+    },
+
+    // Helper for multiple search - for cases when one search term
+    // include another search term (for example, "tamed tame").
+    // Looking for the same elements in the 'matches' and
+    // leave elements with a longer match-length.
+    // Example - before:
+    // matches       1 2 2 2 5 7
+    // matchesLength 4 4 5 6 3 2
+    // After:
+    // matches       1 2 5 7
+    // matchesLength 4 6 3 2
+    clearMatches: function PDFFindController_clearMatches(
+        matchesWithLength, matches, matchesLength) {
+      var i, len;
+      this.sortMatches(matchesWithLength);
+      for (i = 0, len = matchesWithLength.length; i < len; i++) {
+        if (matchesWithLength[i + 1] !== undefined &&
+            matchesWithLength[i].match === matchesWithLength[i + 1].match) {
+          continue;
+        }
+        matches.push(matchesWithLength[i].match);
+        matchesLength.push(matchesWithLength[i].matchLength);
+      }
+    },
+
     calcFindMatch: function PDFFindController_calcFindMatch(pageIndex) {
       var pageContent = this.normalize(this.pageContents[pageIndex]);
       var query = this.normalize(this.state.query);
       var caseSensitive = this.state.caseSensitive;
+      var phraseSearch = this.state.phraseSearch;
       var queryLen = query.length;
+      var matches = [], matchesLength = [], matchesWithLength;
+
+      function addMatches(query) {
+        if (query === '') {
+          return;
+        }
+        var queryLen = query.length;
+        var matchIdx = -queryLen;
+        while (true) {
+          matchIdx = pageContent.indexOf(query, matchIdx + queryLen);
+          if (matchIdx === -1) {
+            return;
+          }
+          if (!phraseSearch) {
+            matchesWithLength.push({ match: matchIdx, matchLength: queryLen });
+          } else {
+            matches.push(matchIdx);
+            matchesLength.push(queryLen);
+          }
```

Comment:
Looking at this further, I am not too concerned about the memory wasted by the repeated values. However, this does seem to be a bit more complicated than necessary.  For example, we could probably do away with both `matches` and `matchesLength` and just have the array `matchesWithLength`. That will use more memory when `phraseSearch==true`, but it won't be too bad, and it will keep the code clean.

Alternatively, if people are concerned about memory, when `phraseSearch==true`, we can set `matchesLength` either to `null` or to `queryLen`, rather than an array of repeated `queryLen`